### PR TITLE
MonoComonad instances for `NonNull`

### PR DESCRIPTION
**Pull request by mstksg**
As discussed in #74 .

Implemented in the best way I could think of; because it uses `tailEx`, requires an `IsSequence` constraint, which I think makes sense.  The implementation might be considered an abuse of 'MonoTraversable', but I think it might be the best way to get the desired result.

By the way, while testing this, I discovered that you can't sanely use any infinite list as a `NonNull`, because it wants to calculate the entire length of the list before declaring that the length is at least 1.  Is this intended behavior?

### CODE REVIEW

**Comment by mstksg** [comment]
By the way, I also noticed that I couldn't directly write `oextract` on the `MinLen`, and that I had to do the newtype unwrapping, because there's no `IsSequence mono => MinLen n mono`.  Is this an intentional design choice? I can submit a PR if not.

File: src/Data/MinLen.hs [hunk 1]
```diff
@@ -480,3 +481,13 @@ minimumBy :: MonoFoldable mono
           -> Element mono
 minimumBy cmp = minimumByEx cmp . unMinLen
 {-# INLINE minimumBy #-}
+
+-- MonoComonad instance for IsSequence mono => NonNull mono
+instance IsSequence mono => MonoComonad (MinLen (Succ Zero) mono) where
```
Comment (snoyberg) [inline src/Data/MinLen.hs]: Instead of `Succ Zero`, `Succ a` would allow this to work for arbitrarily sized `MinLen`s.

**Comment by snoyberg** [comment]
Thanks for the reminder on the infinite list thing. There was a FIXME I completely forgot about, but I've pushed a commit now that should fix it. I've also replaced `Succ Zero` with `Succ a`.

IIRC, we can't implement `IsSequence` for `MinLen` because operations like `filter` may fail (you end up with a possibly smaller container afterwards).

**Comment by snoyberg** [comment]
I think I see now why you went for `Succ Zero` instead of `Succ a`...

**Comment by snoyberg** [comment]
Here's a demonstration of this being broken:

```haskell
import Data.MinLen
import Data.MonoTraversable
import Prelude hiding (head)

main :: IO ()
main = case toMinLen [1,2,3] :: Maybe (MinLen (Succ (Succ Zero)) [Int]) of
    Just ml -> print $ oextend f ml

f :: MinLen (Succ (Succ a)) [Int] -> Int
f = head . tailML
```

Results in:

> MinLen {unMinLen = [2,3,foo.hs: Data.MonoTraversable.headEx: empty

I've fixed up the instance, mind looking this over before I release?

**Comment by mstksg** [comment]
I was only doing it because I was unwrapping the type synonym for `NonNull`, heh.  But the `Succ a` should unify with `NonNull`, shouldn't it?

**Comment by mstksg** [comment]
Oh, didn't see the reply on the demonstration.  Yeah, this definitely comes from being `Succ a` instead of `Succ Zero`, because the definition of `oextend` relies on `f` being able to work on any container from size 1 to `N`.

**Comment by mstksg** [comment]
I have a feeling that you can't write a valid monocomonad if the container must always be of length 2 or more.  I'll think a little bit on it.

**Comment by mstksg** [comment]
Okay, I actually found what I think is a valid comonad instance for containers where you are guaranteed a minimum length.  I haven't proven it on paper but the quickcheck tests seem to work out...and quickcheck proos are proofs, right? :)  (just kidding, I'll look into these more) There are two:

1. A generalization of the current instance, which would most likely have to be defined via induction from a base case.  Still not 100% sure it can be done but I have a feeling it can.

2. One that works for *all* `IsSequence`'s, which is a complete departure from the current instance in terms of the implementation wrt `NonNull`...so it might be unintuitive if you're expecting the behavior from the `NonEmpty`, `ViewR` etc. instances.  It also is probably much less performant because it requires cyclic vector shifting/'rotation'.

(1) probably makes more sense to me, and I can try to implement that.  We wouldn't have to modify any of the existing code either; adding the new instances is just a matter of adding the instances for the inductive cases.  However, the instance is a little...weird, and non-conventional.  I haven't seen it anywhere else.  I don't think anyone would miss it *too* much if it weren't there.  It might still be a good addition though just for completion's sake.

I'll keep you updated on if I can get it working.

**Pull request merged**